File: 0560_subarray_sum_equals_k/subarray_sum.js
/**
 * Time: O(n)
 * Space: O(n)
 */
const subarraySum = (nums, k) => {
  const cache = { 0: 1 }
  return nums.reduce(
    (obj, curr) => {
      obj.total += curr
      const subtotal = obj.total
      obj.res += cache[subtotal - k] || 0
      cache[subtotal] = ++cache[subtotal] || 1
      return obj
    },
    { total: 0, res: 0 }
  ).res
}

File: 0560_subarray_sum_equals_k/subarray_sum.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func subarraySum(nums []int, k int) int {
	ht := make(map[int]int)
	ht[0] = 1
	var total, res int
	for i := 0; i < len(nums); i++ {
		total += nums[i]
		val, ok := ht[total-k]
		if ok {
			res += val
		} else {
			res += 0
		}
		ht[total]++
	}
	return res
}

func main() {

}
